File: typescript/core/test/executor.test.ts
/**
 * Unit tests for the durable-execution seam (ADR-030), mirroring the Rust
 * reference's `executor.rs` / `activities.rs` tests one-for-one so the parity is
 * visible: the in-process executor is a verbatim delegation to the agent's own
 * run entry points, and `driveTurn` reproduces the loop's decision flow.
 */

import { describe, expect, it } from 'vitest';
import { SmoothAgent } from '../src/agent.js';
import type { StreamEvent, Tool } from '../src/agent.js';
import { driveTurn, InProcessActivities, InProcessExecutor } from '../src/executor.js';
import { MockLlmProvider } from '../src/llmProvider.js';

/** Minimal tool that echoes its `text` argument — mirrors the Rust tests' `EchoTool`. */
const echoTool: Tool = {
    name: 'echo',
    description: 'Echoes input back',
    parameters: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] },
    async execute(args) {
        return String(args.text ?? '');
    },
};

/** The seed state `SmoothAgent.run` holds when it enters its loop. */
function seedMessages(user: string): Array<Record<string, unknown>> {
    return [
        { role: 'system', content: 'You are a test agent' },
        { role: 'user', content: user },
    ];
}

describe('InProcessExecutor', () => {
    it('produces the same result as calling the agent run directly', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('the answer is 42');
        const agent = new SmoothAgent(mock, { instructions: 'You are a test agent' });

        const result = await new InProcessExecutor().execute(agent, 'what is the answer?');

        expect(result.text).toBe('the answer is 42');
        expect(result.iterations).toBe(1);
        expect(mock.callCount).toBe(1);
        expect(JSON.stringify(mock.calls[0].messages)).toContain('what is the answer?');
    });

    it('streaming emits events and returns the same final result', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('streamed reply');
        const agent = new SmoothAgent(mock, { instructions: 'You are a test agent' });

        const events: StreamEvent[] = [];
        for await (const event of new InProcessExecutor().executeStreaming(agent, 'stream please')) {
            events.push(event);
        }

        expect(events.length).toBeGreaterThan(1);
        expect(events.filter((e) => e.type === 'text').length).toBeGreaterThan(0);
        const done = events[events.length - 1];
        expect(done.type).toBe('done');
        expect(done.type === 'done' && done.response.text).toBe('streamed reply');
    });
});

describe('driveTurn', () => {
    it('stops after exactly one model call on a plain text reply', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('the answer is 42');
        const activities = new InProcessActivities(mock);

        const messages = seedMessages('what is the answer?');
        await driveTurn(activities, messages, []);

        expect(mock.callCount).toBe(1);
        expect(messages[messages.length - 1]).toEqual({ role: 'assistant', content: 'the answer is 42' });
    });

    it('runs the tool, appends the result paired to the call, and loops back to the model', async () => {
        const mock = new MockLlmProvider();
        mock.pushToolCall('call-1', 'echo', JSON.stringify({ text: 'hello tools' }));
        mock.pushText('done');
        const activities = new InProcessActivities(mock, [echoTool]);

        const messages = seedMessages('use the echo tool');
        await driveTurn(activities, messages, []);

        expect(mock.callCount).toBe(2);
        const toolMsgs = messages.filter((m) => m.role === 'tool');
        expect(toolMsgs).toEqual([{ role: 'tool', tool_call_id: 'call-1', name: 'echo', content: 'hello tools' }]);
        expect(messages[messages.length - 1]).toEqual({ role: 'assistant', content: 'done' });
    });

    it('bounds the loop at maxIterations', async () => {
        const mock = new MockLlmProvider();
        for (let i = 0; i < 5; i++) mock.pushToolCall(`call-${i}`, 'echo', JSON.stringify({ text: 'again' }));
        const activities = new InProcessActivities(mock, [echoTool]);

        const messages = seedMessages('loop forever');
        await driveTurn(activities, messages, [], { maxIterations: 2 });

        expect(mock.callCount).toBe(2);
        expect(messages.filter((m) => m.role === 'tool')).toHaveLength(2);
    });
});

File: go/core/executor_test.go
package core

import (
	"context"
	"testing"
)

// echoTool (tool_hook_test.go) is the Go sibling of the Rust EchoTool these
// tests mirror.
func seedConversation(user string) []ChatMessage {
	return []ChatMessage{
		{Role: "system", Content: "You are a test agent"},
		{Role: "user", Content: user},
	}
}

// The in-process executor drives the loop identically to SmoothAgent.Run: a
// single text response with no tool calls ends the turn after one model call,
// with the assistant content the mock scripted.
func TestInProcessExecutorMatchesAgentRun(t *testing.T) {
	mock := NewMockLlmProvider().PushText("the answer is 42")
	agent := NewSmoothAgent(mock, AgentOptions{Instructions: "be helpful"})

	res, err := NewInProcessExecutor().Execute(context.Background(), agent, "what is the answer?", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "the answer is 42" || res.Iterations != 1 || res.ToolCalls != 0 {
		t.Fatalf("unexpected result: %+v", res)
	}
	if mock.CallCount() != 1 {
		t.Fatalf("want 1 model call, got %d", mock.CallCount())
	}
	found := false
	for _, m := range mock.Calls()[0].Messages {
		if m.Role == "user" && m.Content == "what is the answer?" {
			found = true
		}
	}
	if !found {
		t.Fatalf("user message not sent; messages=%+v", mock.Calls()[0].Messages)
	}
}

// The streaming entry point surfaces events and returns the same final result.
func TestInProcessExecutorStreamingEmitsEventsAndReturnsResult(t *testing.T) {
	mock := NewMockLlmProvider().PushText("streamed reply")
	agent := NewSmoothAgent(mock, AgentOptions{Instructions: "be helpful"})

	stream, err := NewInProcessExecutor().ExecuteStreaming(context.Background(), agent, "stream please", nil)
	if err != nil {
		t.Fatal(err)
	}
	var events []StreamEvent
	var final AgentRunResponse
	for ev := range stream.Events() {
		events = append(events, ev)
		if ev.Kind == StreamDone {
			final = ev.Response
		}
	}
	if err := stream.Err(); err != nil {
		t.Fatal(err)
	}
	if len(events) == 0 {
		t.Fatal("expected streaming events to be emitted")
	}
	if final.Text != "streamed reply" || final.Iterations != 1 {
		t.Fatalf("unexpected final response: %+v", final)
	}
}

// The executor is usable through the interface, which is what lets a consumer
// hold an AgentExecutor and swap backends (Go's analogue of Rust's
// object-safety test).
func TestExecutorIsUsableThroughInterface(t *testing.T) {
	mock := NewMockLlmProvider().PushText("dyn ok")
	agent := NewSmoothAgent(mock, AgentOptions{})

	var executor AgentExecutor = NewInProcessExecutor()
	res, err := executor.Execute(context.Background(), agent, "via interface", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "dyn ok" {
		t.Fatalf("unexpected result: %+v", res)
	}
}

// A plain text reply ends the turn after exactly one model call, with the
// assistant content the mock scripted — no tools run.
func TestDriveTurnTextReplyStopsAfterOneModelCall(t *testing.T) {
	mock := NewMockLlmProvider().PushText("the answer is 42")
	activities := NewInProcessActivities(mock, "test-model", nil)

	convo, err := DriveTurn(context.Background(), activities, seedConversation("what is the answer?"), nil, DefaultTurnPolicy())
	if err != nil {
		t.Fatal(err)
	}
	if mock.CallCount() != 1 {
		t.Fatalf("want 1 model call, got %d", mock.CallCount())
	}
	last := convo[len(convo)-1]
	if last.Role != "assistant" || last.Content != "the answer is 42" {
		t.Fatalf("unexpected tail message: %+v", last)
	}
}

// A tool call is executed and its result appended paired to the call, then the
// follow-up text reply ends the turn — two model calls, one tool result between.
func TestDriveTurnRunsToolThenFinishes(t *testing.T) {
	mock := NewMockLlmProvider()
	mock.PushToolCall("call-1", "echo", `{"text": "hello tools"}`).PushText("done")
	activities := NewInProcessActivities(mock, "test-model", []Tool{echoTool()})

	convo, err := DriveTurn(context.Background(), activities, seedConversation("use the echo tool"), nil, DefaultTurnPolicy())
	if err != nil {
		t.Fatal(err)
	}
	if mock.CallCount() != 2 {
		t.Fatalf("want 2 model calls, got %d", mock.CallCount())
	}
	var toolMsgs []ChatMessage
	for _, m := range convo {
		if m.Role == "tool" {
			toolMsgs = append(toolMsgs, m)
		}
	}
	if len(toolMsgs) != 1 {
		t.Fatalf("want 1 tool message, got %d (%+v)", len(toolMsgs), convo)
	}
	if toolMsgs[0].Content != "hello tools" || toolMsgs[0].ToolCallID != "call-1" {
		t.Fatalf("tool result not paired to the call: %+v", toolMsgs[0])
	}
	// The tool result was fed back to the model on the second call.
	fed := false
	for _, m := range mock.Calls()[1].Messages {
		if m.Role == "tool" && m.Content == "hello tools" {
			fed = true
		}
	}
	if !fed {
		t.Fatalf("tool result not fed back; messages=%+v", mock.Calls()[1].Messages)
	}
	last := convo[len(convo)-1]
	if last.Role != "assistant" || last.Content != "done" {
		t.Fatalf("unexpected tail message: %+v", last)
	}
}

// MaxIterations bounds the loop: a model that keeps requesting tools stops after
// exactly that many model calls, returning what it has rather than an error.
func TestDriveTurnMaxIterationsBoundsLoop(t *testing.T) {
	mock := NewMockLlmProvider()
	for i := 0; i < 5; i++ {
		mock.PushToolCall("call-1", "echo", `{"text": "again"}`)
	}
	activities := NewInProcessActivities(mock, "test-model", []Tool{echoTool()})

	convo, err := DriveTurn(context.Background(), activities, seedConversation("loop forever"), nil, TurnPolicy{MaxIterations: 2})
	if err != nil {
		t.Fatal(err)
	}
	if mock.CallCount() != 2 {
		t.Fatalf("want the loop bounded at 2 model calls, got %d", mock.CallCount())
	}
	// 2 seed + 2 * (assistant + tool result).
	if len(convo) != 6 {
		t.Fatalf("want 6 messages after 2 bounded iterations, got %d (%+v)", len(convo), convo)
	}
}
